File: euxis-bridge/tests/unit/test_policy.py
"""Tests for SkillExecutionPolicy and sub-policies."""

from euxis_bridge.core.policy import (
    FilesystemPolicy,
    NetworkPolicy,
    ResourceLimits,
    SkillExecutionPolicy,
)


def test_default_filesystem_policy() -> None:
    p = FilesystemPolicy()
    assert p.read_only is True
    assert p.write_paths == ()


def test_default_network_policy() -> None:
    p = NetworkPolicy()
    assert p.deny_all is True
    assert p.allowed_hosts == ()


def test_default_resource_limits() -> None:
    r = ResourceLimits()
    assert r.memory_mb == 256
    assert r.timeout_seconds == 20


def test_default_skill_execution_policy() -> None:
    p = SkillExecutionPolicy()
    assert p.require_signature is False
    assert p.public_key_path is None
    assert p.filesystem.read_only is True
    assert p.network.deny_all is True
    assert p.resources.memory_mb == 256
    assert p.output_schema is None
    assert p.audit_log_path is None


def test_permissive_factory() -> None:
    p = SkillExecutionPolicy.permissive()
    assert p.require_signature is False
    assert p.filesystem.read_only is False
    assert p.network.deny_all is False
    assert p.resources.memory_mb == 512
    assert p.resources.timeout_seconds == 60


def test_frozen_immutability() -> None:
    p = SkillExecutionPolicy()
    try:
        p.require_signature = True  # type: ignore[misc]
        raise AssertionError("Expected FrozenInstanceError")
    except AttributeError:
        pass


def test_custom_policy() -> None:
    p = SkillExecutionPolicy(
        require_signature=True,
        public_key_path="/keys/pub.pem",
        filesystem=FilesystemPolicy(read_only=False, write_paths=("/tmp",)),
        network=NetworkPolicy(deny_all=False, allowed_hosts=("api.example.com",)),
        resources=ResourceLimits(memory_mb=128, timeout_seconds=5),
        output_schema={"required": ["result"]},
        audit_log_path="/var/log/audit.jsonl",
    )
    assert p.require_signature is True
    assert p.public_key_path == "/keys/pub.pem"
    assert p.filesystem.write_paths == ("/tmp",)
    assert p.network.allowed_hosts == ("api.example.com",)
    assert p.resources.memory_mb == 128
    assert p.output_schema == {"required": ["result"]}
    assert p.audit_log_path == "/var/log/audit.jsonl"

File: euxis-cpp/euxis-bridge-cpp/tests/test_policy.cpp
#include <gtest/gtest.h>
#include <euxis/bridge/policy.hpp>

namespace euxis::bridge {

TEST(PolicyTest, DefaultValues) {
    SkillExecutionPolicy policy;
    EXPECT_FALSE(policy.require_signature);
    EXPECT_TRUE(policy.filesystem.read_only);
    EXPECT_TRUE(policy.network.deny_all);
    EXPECT_EQ(policy.resources.memory_mb, 256u);
    EXPECT_EQ(policy.resources.timeout_seconds, 20u);
}

TEST(PolicyTest, Permissive) {
    auto policy = SkillExecutionPolicy::permissive();
    EXPECT_FALSE(policy.require_signature);
    EXPECT_FALSE(policy.filesystem.read_only);
    EXPECT_FALSE(policy.network.deny_all);
    EXPECT_EQ(policy.resources.memory_mb, 512u);
    EXPECT_EQ(policy.resources.timeout_seconds, 60u);
}

TEST(PolicyTest, ToJson) {
    SkillExecutionPolicy policy;
    policy.require_signature = true;
    policy.network.allowed_hosts = {"example.com"};

    auto j = policy.to_json();
    EXPECT_TRUE(j["require_signature"]);
    EXPECT_TRUE(j["filesystem"]["read_only"]);
    EXPECT_TRUE(j["network"]["deny_all"]);
    EXPECT_EQ(j["network"]["allowed_hosts"][0], "example.com");
}

TEST(PolicyTest, ResourceLimits) {
    ResourceLimits limits;
    limits.memory_mb = 1024;
    limits.timeout_seconds = 120;
    EXPECT_EQ(limits.memory_mb, 1024u);
    EXPECT_EQ(limits.timeout_seconds, 120u);
}

TEST(PolicyTest, FilesystemPolicy) {
    FilesystemPolicy fs;
    fs.read_only = false;
    fs.write_paths = {"/tmp", "/var/data"};
    EXPECT_FALSE(fs.read_only);
    EXPECT_EQ(fs.write_paths.size(), 2);
}

// --- Coverage: line 19 (to_json with public_key_path set) ---
TEST(PolicyTest, ToJsonWithPublicKeyPath) {
    SkillExecutionPolicy policy;
    policy.public_key_path = "/tmp/pubkey.pem";

    auto j = policy.to_json();
    EXPECT_TRUE(j.contains("public_key_path"));
    EXPECT_EQ(j["public_key_path"], "/tmp/pubkey.pem");
}

// --- Coverage: line 24 (to_json with write_paths) ---
TEST(PolicyTest, ToJsonWithWritePaths) {
    SkillExecutionPolicy policy;
    policy.filesystem.read_only = false;
    policy.filesystem.write_paths = {"/tmp", "/var/data"};

    auto j = policy.to_json();
    EXPECT_EQ(j["filesystem"]["write_paths"].size(), 2u);
    EXPECT_EQ(j["filesystem"]["write_paths"][0], "/tmp");
}

// --- Coverage: line 32 (to_json with output_schema set) ---
TEST(PolicyTest, ToJsonWithOutputSchema) {
    SkillExecutionPolicy policy;
    policy.output_schema = nlohmann::json{{"type", "object"}};

    auto j = policy.to_json();
    EXPECT_TRUE(j.contains("output_schema"));
    EXPECT_EQ(j["output_schema"]["type"], "object");
}

// --- Coverage: line 35 (to_json with audit_log_path set) ---
TEST(PolicyTest, ToJsonWithAuditLogPath) {
    SkillExecutionPolicy policy;
    policy.audit_log_path = "/var/log/audit.jsonl";

    auto j = policy.to_json();
    EXPECT_TRUE(j.contains("audit_log_path"));
    EXPECT_EQ(j["audit_log_path"], "/var/log/audit.jsonl");
}

}  // namespace euxis::bridge
